jupyterhub_traefik_proxy/performance/proxy_perf.py

```diff
@@ -0,0 +1,294 @@
+        formatter_class=argparse.RawTextHelpFormatter,
+    )
+
+    parser.add_argument(
+        "--measure",
+        dest="metric",
+        default="methods",
+        help=textwrap.dedent(
+            """\
+            What metric to measure. Available metrics:
+            - methods
+            - http_throughput_small
+            - http_throughput_large
+            - ws_throughput
+            If no metric is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--proxy",
+        dest="proxy_class",
+        default="TomlProxy",
+        help=textwrap.dedent(
+            """\
+            Proxy class to analyze.
+            Available proxies:
+            -TomlProxy
+            -EtcdProxy
+            -CHP
+            If no proxy is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--routes_number",
+        dest="routes_number",
+        default=10,
+        help=textwrap.dedent(
+            """\
+            Number of routes to be added/deleted/retrieved.
+            If no number is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--concurrent_requests_number",
+        dest="concurrent_requests_number",
+        default=10,
+        help=textwrap.dedent(
+            """\
+            Number of concurrent requests when comoputing the throughput.
+            If no number is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--output",
+        dest="csv_filename",
+        help=textwrap.dedent(
+            """
+            The csv file name where the results will be stored.
+            If no file is provided, the results will only be outputed to stdout.
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--backend_port",
+        dest="backend_port",
+        default=9000,
+        help=textwrap.dedent(
+            """
+            The port where the backend receives http/websocket requests.
+            If no port is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    args = parser.parse_args()
+    metric = args.metric
+    proxy_class = args.proxy_class
+    routes_number = int(args.routes_number)
+    concurrent_requests_number = int(args.concurrent_requests_number)
+    csv_filename = args.csv_filename
+    backend_port = args.backend_port
+
+    requests_no = 1000
+
+    loop = asyncio.get_event_loop()
+    if metric == "methods":
+        result = loop.run_until_complete(
+            measure_methods_performance(proxy_class, routes_number)
+        )
+        if csv_filename:
+            with open(csv_filename, mode="w+") as csv_file:
+                fieldnames = [("T_" + str(i)) for i in range(routes_number)]
```

Comment:
Maybe don't need to sample every single point in 1...1000. It could be appropriate to do a variable-density sampling, e.g.

```python
def staggered_sample(N):
    samples = []
    upper_limit = 10
    stride = 1
    previous = 1
    while True:
        samples.extend(range(previous, min(upper_limit, N+1), stride))
        if upper_limit >= N:
            return samples
        previous = upper_limit
        upper_limit *= 2
        stride *= 2
```

or sample in logspace, given the number of points you want to collect, e.g.

```python
N = 1000
np.unique(np.logspace(0, np.log10(N), 50, dtype=int))
```

for 50 points distributed evenly on a log-scale from 1 to 1000.



It might also be useful to collect some statistics by running each test a few times. The [timeit module](https://docs.python.org/3/library/timeit.html) gives some handy utilities for running samples.

You could either record raw samples, or just the summary mean, min, max for error bars.

It may be useful to collect this data as bigger CSVs with the number of routes as a number instead of a column title, e.g.

```csv
proxy, method, nroutes, t
toml, get_routes, 10, 1.125
```

Then you can do things like use pandas groupby to make plots of different data vs each other, e.g. "show me get_routes with 100 routes for each proxy" or "show me get_routes for toml against the number of routes" etc.

Reply:
@minrk, I've realized I wasn't very descriptive in the comments and I want to clarify some things.
When I measure how much time it takes to add 1 to 500 routes to the routing table, I'm measuring how much it takes to add one route when the routing table contains
* 0 routes
* 1 route
....
* 499 routes
Same when I delete them, but in reverse order. I measure how much it takes to delete a route when the routing table contains:
* 500 routes
* 499 routes
....
* 1 route
As for the get_all_routes, I'm calling it 500 times and compute the mean time.

Unfortunately I messed up the plot of the *delete_route* method for all the proxies because I should have labeled the time taken in reverse order (T_500, .. T_0 on the plot). So TomlProxy is slower when it has to delete a route from a routing table which has a large number of routes. This makes sense since we're accessing the disk to persist the routing table after each delete.

> adding routes via etcd appears to scale linearly with the number of routes

This is indeed worrisome. I have an intuition that this might happen because we're calling *add_route* 500 times in a for loop really fast and because adding the routes in etcd might take slightly longer, the tasks get queued in the executor thread 
of the proxy (that we used to make the transactions async). So basically the last call to add_route (for the 500th route) might wait for some older calls to finish.
I'll investigate if this is actually what's happening.